Migrate operator to operator-sdk v1.19.2

supreeth7 opened the pull request:
### What type of PR is this? 
Refactor

### What this PR does / why we need it:
- With this PR we are upgrading operator-sdk from v0.17.2 to v1.19.2.
- This involves a lot of breaking changes as the latest o-sdk has a new directory structure and updated dependencies.
- A detailed explanation on how the migration was done can be found [here](https://github.com/openshift/ops-sop/blob/master/operator_sdk/migrate_operator_sdk.md).

**Notable changes:-**
1. No direct dependency on operator-sdk. We now use `controller-gen` to generate manifests.

2. `listType=atomic` annotation added to the API types to support new openapi-gen version.

3. Removed `cluster-api` dependency. The two functions for which the library was being used have been moved to `pkg/util/util.go`.

4. `runtime.Object` is now `client.Object`, compatible with newer version of `controller-runtime`.

5. FIPS support.

### Which issue(s) this PR fixes(can be a reference to existing issue or jira/bz):
Resolves: [OSD-10739](https://issues.redhat.com/browse/OSD-10739)

### Is it a breaking change or backward compatible?
Breaking

supreeth7 commented:
[#220](https://github.com/openshift/boilerplate/pull/220) has been created to fix the failing prow test.

supreeth7 commented:
Validation has been fixed. Now onto test and coverage. 😠

georgettica commented:
Maybe you need to set the GOCACHE?

supreeth7 commented:
> Maybe you need to set the GOCACHE?

Yes, after all the tooling has been fixed the only error that both the failing tests point to are related to creating `/. cache` directory. My guess is images running on prow, like the test job don't have root permissions. Probably need to point GOCACHE  to `/tmp` directory which would have write permissions.

supreeth7 commented:
Finally all tests pass, phew.
This is now ready to be reviewed.

supreeth7 commented:
/label tide/merge-method-squash

georgettica commented on Makefile:
```diff
@@ -1,7 +1,5 @@
 include boilerplate/generated-includes.mk
 
-SHELL := /usr/bin/env bash
```
not complaining, but why?

georgettica replied:
That is put Makefile, but ok

georgettica commented on build/Dockerfile:
```diff
@@ -11,4 +13,5 @@ ENV OPERATOR_PATH=/gcp-project-operator \
 COPY --from=builder /gcp-project-operator/build/_output/bin/${OPERATOR_BIN} /usr/local/bin/${OPERATOR_BIN}
 LABEL io.openshift.managed.name="gcp-project-operator" \
-      io.openshift.managed.description="This operator will be responsible for managing GCP project and credentials."
+    io.openshift.managed.description="This operator will be responsible for managing GCP project and credentials."
```
is that better whitespace wize? or was that by your editor

if it was by your editor, we should card later (for boilerplate) to lint the dockerfiles for a standard format so we don't need to review on this even

supreeth7 replied:
It was my editor (VScode).
Lint checks on Dockerfiles sounds good to me, will have it carded after discussing it with FL.

georgettica commented on config/templates/csv-template.yaml:
```diff
@@ -6,40 +6,40 @@ metadata:
+        - name: gcp-project-operator
+          # Deployment spec will be added here by the generate-operator-bundle.py script.
   customresourcedefinitions:
     owned:
     # CRD's will be added here by the generate-operator-bundle.py
```
ran a `yq -P -i` and the comment was removed.
could you switch the lines so the next `yq -P` won't show a diff? (or however you linted this yaml)

created an issue for YQ with this example https://github.com/mikefarah/yq/issues/1217

supreeth7 replied:
Sorry I didn't get you, right `yq -P -i <file>` does remove the last line. Do you want it removed or do you want me to switch lines _44_ & _45_?

``` 
    # CRD's will be added here by the generate-operator-bundle.py
    owned:
```

georgettica replied:
Switch them so for now it'll not be removed

georgettica commented on controllers/projectclaim_controller.go:
```diff
@@ -0,0 +1,130 @@
+limitations under the License.
+*/
+
+package controllers
```
doesn't this mean that the 2 controllers can share data?
they are supposed to be separate AFAICT
same goes to the tests

supreeth7 replied:
The same thought crossed my head too, seeing the two controllers were separate pre-migration.
Operator SDK by default generated all the controllers under the same folder and I kept it that way and wanted the reviewers to make a call on this.
Let me try separating them now...

supreeth7 replied:
Separation done.

georgettica commented on controllers/projectclaim_controller_test.go:
```diff
@@ -1,38 +1,39 @@
-			mockClient,
-			scheme.Scheme,
-		)
+		reconciler = &ProjectClaimReconciler{
+			Client: mockClient,
+			Scheme: scheme.Scheme,
+		}
```
good improvement, +1!

georgettica commented on controllers/projectclaimadapter.go:
```diff
@@ -87,9 +84,9 @@ func (c *ProjectClaimAdapter) ProjectReferenceExists() (bool, error) {
-func (adapter *ProjectClaimAdapter) EnsureProjectClaimDeletionProcessed() (gcputil.OperationResult, error) {
-	if adapter.IsProjectClaimDeletion() {
-		crState, err := adapter.FinalizeProjectClaim()
+func (c *ProjectClaimAdapter) EnsureProjectClaimDeletionProcessed() (gcputil.OperationResult, error) {
```
this is not supposed to be c, but p.
nitpicking as CAD has a linter that checks this for us (I think `revive`)

supreeth7 replied:
I've kept it `c` for now.

georgettica commented on pkg/util/mocks/structs/core.go:
```diff
@@ -22,8 +22,9 @@ func NewTestSecretBuilder(secretName, namespace, creds string) *testSecretBuilde
-				Namespace: namespace,
+				Name:            secretName,
+				Namespace:       namespace,
+				ResourceVersion: "999",
```
add a comment why is this needed?

georgettica commented on deploy/operator.yaml:
```diff
@@ -18,7 +18,7 @@ spec:
           image: quay.io/app-sre/gcp-project-operator
           command:
-          - gcp-project-operator
+            - gcp-project-operator
```
also this file can be fixed a bit by the `yq -P -i`

georgettica commented on controllers/projectreference_controller_test.go:
```diff
@@ -88,7 +84,7 @@ parentFolderID: fake-folder
 		})
 		It("Returns without error", func() {
-			_, err := reconciler.Reconcile(reconcile.Request{NamespacedName: projectReferenceName})
+			_, err := reconciler.Reconcile(context.TODO(), reconcile.Request{NamespacedName: projectReferenceName})
```
add a global `context.TODO()` maybe?

georgettica commented on controllers/projectreference_adapter.go:
```diff
@@ -538,7 +529,7 @@ func (r *ReferenceAdapter) configureBillingAPI() error {
 	}
 
-	if !util.Contains(enabledAPIs, "cloudbilling.googleapis.com") {
+	if !gcputil.Contains(enabledAPIs, "cloudbilling.googleapis.com") {
```
wasn't is called it `util` in the other files?

supreeth7 replied:
Right, I found a lot of pkg naming inconsistencies along the way and changed them to my best extent.
Changing this to `util`.

georgettica commented on controllers/projectclaimadapter_test.go:
```diff
@@ -1,28 +1,27 @@
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	. "github.com/openshift/gcp-project-operator/controllers"
```
do we need to do this dot import for `	. "github.com/openshift/gcp-project-operator/controllers"`?

georgettica commented on api/v1alpha1/groupversion_info.go:
```diff
@@ -0,0 +1,36 @@
+
+var (
+	// GroupVersion is group version used to register these objects
+	GroupVersion = schema.GroupVersion{Group: "gcp.managed.openshift.io", Version: "v1alpha1"}
```
TODO: we might want to move this to `v1` as the schema didn't change in at least a year and we are using it in prod for a long while

supreeth7 replied:
This would be another major change. Can we keep that for another PR if possible?

georgettica reviewed (commented):
tons of comments. most don't block you from merging.

feel free to contest all and if this works on a CRC. I would rather solve things on stage than keep this massive PR open for longer

georgettica commented:
@supreeth7 I gave lots of comments. I hope it's not overwhelming
please let me know if you were not able to test this

supreeth7 commented:
Thank you for taking your time out and reviewing this @georgettica.
I'll make a few changes addressing your comments.

But, I would need some help in testing it though. You might recall that rather long thread in Slack™ and to this day I still do not have permissions to create folders in GCP. I could use the help of one of the maintainers here as they already have everything in place to test this, would be a straight forward dry-run test I say.

wanghaoran1988 commented on .gitignore:
```diff
@@ -1,90 +1,39 @@
 # Temporary Build Files
```
Could we keep this gitignore files, I see a lot iterms are removed.

supreeth7 replied:
AFAIK those were auto-generated by old o-sdk. The .gitignore generated by the new o-sdk didn't have these lines.

georgettica commented:
/approve
/lgtm

NautiluX commented:
/approve
/hold
unhold to merge when ready

georgettica commented:
@supreeth7 please TAL at https://github.com/mikefarah/yq/issues/1217#event-6696380032 as they solved the issue in a later release, so we can remove the comment / change to the way we see fit
